src/Components/Components/src/Rendering/Renderer.cs

```diff
@@ -68,6 +71,8 @@ public Renderer(IServiceProvider serviceProvider)
         public Renderer(IServiceProvider serviceProvider, IDispatcher dispatcher) : this(serviceProvider)
         {
             _dispatcher = dispatcher;
+            _logger = (ILogger<Renderer>)serviceProvider.GetService(typeof(ILogger<Renderer>))
```

Comment:
Shouldn't this be done in the other ctor? Also how do you feel about passing in an `ILogger` \ `ILoggerFactory` instead to the ctor? This feels a little gross.

Reply:
> Shouldn't this be done in the other ctor?

Yes, absolutely right! I knew to do that earlier then for some reason started working on something else and forgot about it. Thanks for catching that.

> Also how do you feel about passing in an ILogger \ ILoggerFactory instead to the ctor? This feels a little gross. This feels a little gross.

I did this because we already take a `serviceProvider`, so why complicate things for the consumer? Under what circumstances would someone have reason to supply a different `ILogger` than the one the service provider would offer? I'm not absolutely set against it, but would like to make a choice based on practical reasoning.

In the long term, we may need to consume other services too. If we keep adding them as extra constructor overloads one-by-one, it will not be a nice experience for consumers.

Reply:
Who will override the renderer in the first place other than our platforms?

> I did this because we already take a `serviceProvider`, so why complicate things for the consumer

It makes the dependencies visible

> Under what circumstances would someone have reason to supply a different `ILogger` than the one the service provider would offer

Never, but mostly because it is very unlikely to replace the renderer in the stack

Reply:
> n the long term, we may need to consume other services too. If we keep adding them as extra constructor overloads one-by-one, it will not be a nice experience for consumers.

I don't think there is an scenario, other than the platform for consuming the renderer.

Reply:
> Who will override the renderer in the first place other than our platforms?
> I don't think there is an scenario, other than the platform for consuming the renderer.

I don't agree that only people who work for Microsoft will create and/or consume renderers. We have examples of renderers for Electron and Flutter, so it would be surprising if nobody ever had the idea of making or hosting another renderer on a different UI technology, or for use in a different way such as a unit testing system.

> It makes the dependencies visible

Not sure about that.

In terms of *possible* dependencies that may be consumed, the renderer possibly depends on every possible service that an `IServiceProvider` can provide, since it's responsible for instantiating other components that themselves require arbitrary services.

In terms of *required* dependencies, `ILogger<Renderer>` isn't required (it's an optional dependency by the logic here).

Reply:
Thinking about this more, this particular detail just doesn’t really bother me either way. There are two of you saying you prefer something else, so I’ll switch this to be a constructor param.